File: Java/src/se/liu/albhe576/project/Bounds.java
package se.liu.albhe576.project;

import org.lwjgl.BufferUtils;

import java.awt.*;
import java.nio.ByteBuffer;

public class Bounds
{
    private float width;
    private float height;
    private Color color;
    private int thickness;

    private ByteBuffer byteBuffer;

    private float textureOffsetX;
    private float textureOffsetY;

    public ByteBuffer getByteBuffer(){
        return this.byteBuffer;
    }

    public Color getColor(){
        return this.color;
    }

    public int getThickness(){
        return this.thickness;
    }

    public float getWidth(){
        return this.width;
    }
    public float getHeight(){
        return this.height;
    }
    public float getTextureOffsetX(){
        return this.textureOffsetX;
    }
    public float getTextureOffsetY(){
        return this.textureOffsetY;
    }

    public Bounds(float width, float height, float xOffset, float yOffset, Color color, int thickness, int textureWidth, int textureHeight){
        this.width = width;
        this.height = height;
        this.textureOffsetX = xOffset;
        this.textureOffsetY = yOffset;
        this.color = color;
        this.thickness = thickness;
        this.byteBuffer = Bounds.getBoundsBuffer(textureWidth, textureHeight, color, thickness);
    }

    public static ByteBuffer getBoundsBuffer(int width, int height, Color color, int thickness){
        final int bpp = 4;
        int cap = width * height * bpp;

        ByteBuffer byteBuffer = BufferUtils.createByteBuffer(cap);
        for(int i = 0; i < cap; i++){
            byteBuffer.put((byte)0);
        }
        int bottomRow = (height - thickness) * width * bpp;
        for(int i = 0; i < width * thickness * bpp; i+=bpp){
            byteBuffer.put(i + 0, (byte)color.getRed());
            byteBuffer.put(i + 1, (byte)color.getGreen());
            byteBuffer.put(i + 2, (byte)color.getBlue());
            byteBuffer.put(i + 3, (byte) 0xFF);

            byteBuffer.put(bottomRow + i + 0, (byte)color.getRed());
            byteBuffer.put(bottomRow + i + 1, (byte)color.getGreen());
            byteBuffer.put(bottomRow + i + 2, (byte)color.getBlue());
            byteBuffer.put(bottomRow + i + 3, (byte)color.getAlpha());
        }

        int lastColumn = (width - thickness) * bpp;
        for(int i = 0; i < height; i++){
            int yOffset = i * width * bpp;
            for(int j = 0; j < thickness * bpp; j += 4){
                byteBuffer.put(yOffset + j + 0, (byte)color.getRed());
                byteBuffer.put(yOffset + j + 1, (byte)color.getGreen());
                byteBuffer.put(yOffset + j + 2, (byte)color.getBlue());
                byteBuffer.put(yOffset + j + 3, (byte)color.getAlpha());

                byteBuffer.put(yOffset + j + lastColumn + 0, (byte)color.getRed());
                byteBuffer.put(yOffset + j + lastColumn + 1, (byte)color.getGreen());
                byteBuffer.put(yOffset + j + lastColumn + 2, (byte)color.getBlue());
                byteBuffer.put(yOffset + j + lastColumn + 3, (byte)color.getAlpha());
            }
        }


        byteBuffer.flip();

        return byteBuffer;
    }
}

File: C/src/bounds.c
#include "bounds.h"

#if DEBUG_BOUNDS
static void initDebugBounds(struct Image* image, struct Vec3i32 color)
{
  int lineThickness = 6;
  int bpp           = 4;

  image->width      = 128;
  image->height     = 128;
  image->bpp        = bpp;
  i32 size          = image->width * image->height * bpp;
  image->data       = (unsigned char*)malloc(sizeof(unsigned char*) * size);
  for (i32 i = 0; i < size; i++)
  {
    image->data[i] = 0;
  }

  int bottomRow = (image->height - lineThickness) * image->width * bpp;

  for (int i = 0; i < image->width * lineThickness * bpp; i += 4)
  {
    image->data[i + 0]             = color.x;
    image->data[i + 1]             = color.y;
    image->data[i + 2]             = color.z;
    image->data[i + 3]             = 255;

    image->data[bottomRow + i + 0] = color.x;
    image->data[bottomRow + i + 1] = color.y;
    image->data[bottomRow + i + 2] = color.z;
    image->data[bottomRow + i + 3] = 255;
  }

  int lastColumn = (image->width - lineThickness) * bpp;
  for (int i = 0; i < image->height; i++)
  {
    int yOffset = i * image->width * bpp;
    for (int j = 0; j < lineThickness * bpp; j += 4)
    {
      image->data[yOffset + j + 0]              = color.x;
      image->data[yOffset + j + 1]              = color.y;
      image->data[yOffset + j + 2]              = color.z;
      image->data[yOffset + j + 3]              = 255;

      image->data[yOffset + lastColumn + j + 0] = color.x;
      image->data[yOffset + lastColumn + j + 1] = color.y;
      image->data[yOffset + lastColumn + j + 2] = color.z;
      image->data[yOffset + lastColumn + j + 3] = 255;
    }
  }
}

#endif
